restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java

```diff
@@ -1013,221 +930,85 @@ private JsonNode searchWithOnlyInMemoryFilters(
-          }
         }
       } else {
         List<String> segmentsList = Arrays.asList(pathSegmentSplit);
         // left pad any array segments to 6 places
         segmentsList =
             segmentsList.stream().map(this::convertArrayPath).collect(Collectors.toList());
 
-        manyPathsFound = true;
         predicates.add(BuiltCondition.of("p" + i, Predicate.IN, segmentsList));
       }
     }
+  }
 
-    List<FilterCondition> inCassandraFilters = new ArrayList<>();
-    List<FilterCondition> inMemoryFilters = filters;
-
-    if (fields.isEmpty()) {
-      inCassandraFilters =
-          filters.stream()
-              .filter(f -> !FilterOp.LIMITED_SUPPORT_FILTERS.contains(f.getFilterOp()))
-              .collect(Collectors.toList());
-      inMemoryFilters =
-          filters.stream()
-              .filter(f -> FilterOp.LIMITED_SUPPORT_FILTERS.contains(f.getFilterOp()))
-              .collect(Collectors.toList());
-    }
-
-    if ((recurse == null || !recurse)
-        && path.size() < db.MAX_DEPTH
-        && !inCassandraFilters.isEmpty()) {
-      predicates.add(BuiltCondition.of("p" + i++, Predicate.EQ, filters.get(0).getField()));
-    } else if (!path.isEmpty()) {
-      predicates.add(BuiltCondition.of("p" + i++, Predicate.GT, ""));
-    }
+  private List<BuiltCondition> buildConditions(
+      FilterCondition condition, boolean booleanAsNumeric) {
+    List<BuiltCondition> predicates = new ArrayList<>();
+    collectNestedElementConditions(predicates, condition.getPath());
+    collectFieldConditions(predicates, condition);
 
     // The rest of the paths must match empty-string
-    while (i < db.MAX_DEPTH && !path.isEmpty()) {
-      predicates.add(BuiltCondition.of("p" + i++, Predicate.EQ, ""));
-    }
-
-    for (FilterCondition filter : inCassandraFilters) {
-      // All fully supported filters are SingleFilterConditions, as of now
-      SingleFilterCondition singleFilter = (SingleFilterCondition) filter;
-      FilterOp queryOp = singleFilter.getFilterOp();
-      String queryValueField = singleFilter.getValueColumnName();
-      Object queryValue = singleFilter.getValue(db.treatBooleansAsNumeric());
-      if (queryOp != FilterOp.EXISTS) {
-        predicates.add(BuiltCondition.of(queryValueField, queryOp.predicate, queryValue));
-      }
-    }
-
-    ResultSet r;
-
-    predicates.addAll(additionalWhereConditions);
-    if (!predicates.isEmpty()) {
-      r =
-          db.executeSelect(
-              keyspace,
-              collection,
-              predicates,
-              true,
-              paginator.dbPageSize,
-              paginator.getCurrentDbPageState());
-    } else {
-      r =
-          db.executeSelectAll(
-              keyspace, collection, paginator.dbPageSize, paginator.getCurrentDbPageState());
+    for (int i = condition.getPath().size() + 1; i < MAX_DEPTH; i++) {
+      predicates.add(BuiltCondition.of("p" + i, Predicate.EQ, ""));
```

Comment:
I actually think that we don't need to have all fields till the depth empty, but only the first one. I see no situation where first next would be empty and then we would have some on the `next + 1, 2, 3, ...`.

Potentially the query can be faster as it does not have to confirm that values in all 64 `p` columns?

@EricBorczuk Am I wrong here?

Reply:
We need to distinguish the PK for property like `a.b.c` from the PK for `a.b.c.d`. The `p0`, `p1`, `p2` are the same but the `p3` is `<empty>` vs. `d`

Reply:
I guess you mean to set `p3` in that ^ example, but not to set `p4`... `p64`... yes that should work, I guess

Reply:
yes because we only need one predicate then:

```
predicates.add(BuiltCondition.of("p" + condition.getPath().size() + 1, Predicate.EQ, ""));
```

right?

Reply:
but let's do it in a follow-up PR :) I'd like to minimize functional changes here since this PR is already big and is mostly about pagination, not query efficiency. WDYT?

Reply:
Yea, I agree - just to limit the scope